packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1273,7 +1279,7 @@ async function renderToStream(
 
   try {
     // This is a dynamic render. We don't do dynamic tracking because we're not prerendering
-    const RSCPayload = await getRSCPayload(tree, ctx, asNotFound)
+    const RSCPayload = await getRSCPayload(tree, ctx, res.statusCode === 404)
```

Comment:
Is there a way we can just infer it is a 404 from the tree itself?

Reply:
might need more checks since a non-existed (`/404`) page might have slight different tree from the not-found tree we created when it's not-found path. I'll keep it in mind to see if we can optimize it later